source/client-side-operations-timeout/client-side-operations-timeout.rst

```diff
@@ -0,0 +1,866 @@
+includes a ``maxTimeMS`` field and the ``timeoutMS`` option is set. Drivers
+MUST document the behavior of ``runCommand`` for this case and MUST NOT
+attempt to check the command document for the presence of a ``maxTimeMS``
+field.
+
+See `runCommand behavior`_.
+
+Test Plan
+=========
+
+See the `README.rst
+<https://github.com/divjotarora/specifications/blob/csot-tests/source/client-side-operations-timeout/tests/README.rst>`__
+in the tests directory.
+
+Motivation for Change
+=====================
+
+Users have many options to set timeouts for various parts of operation
+execution including, but not limited to, ``serverSelectionTimeoutMS``,
+``socketTimeoutMS``, ``connectTimeoutMS``, ``maxTimeMS``, and ``wTimeoutMS``.
+As a result, users are often unsure which timeout to use. Because some of
+these timeouts are additive, it is difficult to set a combination which
+ensures control will be returned to the user after a specified amount of
+time. As a result, changes are required to the drivers API to deprecate some
+of the existing timeouts and add a new one to specify the maximum execution
+time for an entire operation from start to finish.
+
+Design Rationale
+================
+
+timeoutMS cannot be changed to unset once it’s specified
+--------------------------------------------------------
+
+If ``timeoutMS`` is specified at any level, it cannot be later changed to
+unset at a lower level. For example, a user cannot do:
+
+.. code:: python
+
+   client = MongoClient(uri, timeoutMS=1000)
+   db = client.database("foo", timeoutMS=None)
+
+This is because drivers return existing exception types if ``timeoutMS`` is
+not specified, but will return new exception types and use new timeout
+behaviors if it is. Once the user has opted into this behavior, we should not
+allow them to opt out of it at a lower level. If a user wishes to set the
+timeout to infinite for a specific database, collection, or operation, they
+can explicitly set ``timeoutMS`` to 0.
+
+serverSelectionTimeoutMS is not deprecated
+------------------------------------------
+
+The original goal of the project was to expose a single timeout and deprecate
+all others. This was not possible, however, because executing an operation
+consists of two distinct parts. The first is selecting a server and checkout
+out a connection from its pool. This should have a default timeout because
+failure to do this indicates that the deployment is not in a healthy state or
+that there was a configuration error which prevents the driver from
+successfully connecting. The second is server-side operation execution, which
+cannot have a default timeout. Some operations finish in a few milliseconds,
+while others can run for many hours. Adding a default would inevitably break
+applications. To accomplish both of these goals, ``serverSelectionTimeoutMS``
+was preserved and is used to timeout the client-side section of operation
+execution.
+
+connectTimeoutMS is not deprecated
```

Comment:
Under what circumstances would a user change the value of connectTimeoutMS from its default? It would be really useful to enumerate a few use cases here.

Reply:
The default is 10 seconds, which seems like enough to establish a new connection in almost any scenario. The only use case I'm aware of is to provide a hint about network latency. Maybe in a very high latency environment, a user might want to allow socket establishment to take longer than 10s? Are you aware of any other use cases?

Reply:
I'm not. I've always felt that this option was just noise that confused users more than helped them. Maybe we should require some text in the docs for this option to say something like "The default of 10 seconds should work for most applications"?

Reply:
I think that's a good idea, but I'm not sure where it would go. The new CSOT spec just defines the `timeoutMS` option, but AFAIK some of our existing options like `socketTimeoutMS` and `connectTimeoutMS` aren't defined anywhere except for the URI options spec. I could potentially see them going in this spec as they're related to timeouts or maybe in a new Connections spec that would also define behavior for features like TLS and keep alive.

Given that `connectTimeoutMS` is an existing option and this project is already large, I'd prefer to file a DRIVERS ticket to figure out where it should be defined and take a look at it after this project closes. Does that sound OK?

Reply:
That's fine with me. Open a ticket for us to come up with a better doc string for that option.

Reply:
Filed DRIVERS-1487.